sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/ThroughputRequestThrottler.java

```diff
@@ -44,44 +58,58 @@ private void updateAvailableThroughput() {
     }
 
     public <T> Mono<T> processRequest(RxDocumentServiceRequest request, Mono<T> originalRequestMono) {
-        if (this.availableThroughput.get() > 0) {
-            return originalRequestMono
-                .doOnSuccess(response -> this.trackRequestCharge(response))
-                .doOnError(throwable -> this.trackRequestCharge(throwable));
-        } else {
-            // there is no enough throughput left, block request
-            RequestRateTooLargeException requestRateTooLargeException = new RequestRateTooLargeException();
-
-            int backoffTimeInMilliSeconds = (int)Math.floor(Math.abs(this.availableThroughput.get() * 1000 / this.scheduledThroughput.get()));
-            requestRateTooLargeException.getResponseHeaders().put(
-                HttpConstants.HttpHeaders.RETRY_AFTER_IN_MILLISECONDS,
-                String.valueOf(backoffTimeInMilliSeconds));
-
-            requestRateTooLargeException.getResponseHeaders().put(
-                HttpConstants.HttpHeaders.SUB_STATUS,
-                String.valueOf(HttpConstants.SubStatusCodes.THROUGHPUT_CONTROL_REQUEST_RATE_TOO_LARGE));
-
-            if (request.requestContext != null) {
-                BridgeInternal.setResourceAddress(requestRateTooLargeException, request.requestContext.resourcePhysicalAddress);
+        try {
+            this.throughputReadLock.lock();
```

Comment:
here mixing the readLock pattern with reactor Mono is interesting, and I am not sure if it does what you intended to do.

Think about this, 
1. let's say you acquire the read lock this.throughputReadLock.lock();
2. then we are returning a Mono from the processRequest(.) method. which means as we go out of the scope of this method the finally block and `this.throughputReadLock.unlock();` will be invoked. so now the lock is released.
3. now `trackRequestCharge` and `trackRequestCharge` won't be invoked till we go out of the scope of `processRequest ` and someone subscribes to the returned `Mono`.

I checked the `trackRequestCharge` and `trackRequestCharge` implementation I noticed that you are re-accquiring the read lock.

Now please consider this scenario: does this cause any problem?
1. `processRequest acquires the lock this.throughputReadLock.lock(); and read the value of `availableThroughput`
2. `processRequest  returns a Mono and release the ` this.throughputReadLock.lock();` lock in the finally block
3. now let's say some other thread acquires the write lock and updates the throughputs
4. someone subsctibes to the Mono returned in step 2, and now `trackRequestCharge` and `trackRequestCharge` will be invoked, they will attempt to accquire readLock but because of the step 3. in between the throughput values they see is not the same values that the if branch of `processRequest` observed.

Does this cause any problem?

Reply:
@moderakh  @FabianMeiswinkel  @kushagraThapar 

I think this is one of the consequences of we only track the RU usage based on the response compared to pre-calculation.

The consequences of it : 
   a. could be some requests actually happened in the previous cycle their response will be tracked in the next cycle and eat some throughput from the next cycle. 
   b. The throughput usage for previous cycle is lower than the real usage.

It can be improved if we do some operation RU predication later on. 
Or we can do some stuff like: for each operation, it can be attached to a certain cycle version. and we only track the response if it matches to the current cycle version. Maybe this is a better way?

The reason I introduced the read write lock is mainly to keep ScheduledThroughput and availableThroughput to be in sync, since if we are going to return 429, then we depends on those two values to calculate the backoff time.

Any throughputs for this one?